### PR TITLE
8066859: java/lang/ref/OOMEInReferenceHandler.java failed with java.lang.Exception: Reference Handler thread died

### PR DESCRIPTION
**Pull request by DougLea**
8066859 : java/lang/ref/OOMEInReferenceHandler.java failed with java.lang.Exception: Reference Handler thread died

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8066859](https://bugs.openjdk.org/browse/JDK-8066859): java/lang/ref/OOMEInReferenceHandler.java failed with java.lang.Exception: Reference Handler thread died

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9427/head:pull/9427` \
`$ git checkout pull/9427`

Update a local copy of the PR: \
`$ git checkout pull/9427` \
`$ git pull https://git.openjdk.org/jdk pull/9427/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9427`

View PR using the GUI difftool: \
`$ git pr show -t 9427`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9427.diff">https://git.openjdk.org/jdk/pull/9427.diff</a>

</details>

### CODE REVIEW

**Comment by DougLea** [comment]
Thanks to @AlanBateman for suggesting to disable possibly misleading stack traces in pre-allocated exceptions; now updated

**Comment by DougLea** [comment]
I keep trying but failing to change commit message to something jcheck likes?

**Comment by AlanBateman** [comment]
> I keep trying but failing to change commit message to something jcheck likes?

There should be an "Edit" button in the top right that allows you to edit the title and change it to "8066859 : java/lang/ref/OOMEInReferenceHandler.java failed ...".   There are two Skara commands that may be useful here. 
Adding a comment "/issue JDK-8066859" will link the PR to this JBS issue. There is also a "/summary" command to edit the commit message.

File: src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedLongSynchronizer.java [hunk 1]
```diff
@@ -271,14 +293,17 @@ final int acquire(Node node, long arg, boolean shared,
         Thread current = Thread.currentThread();
         byte spins = 0, postSpins = 0;   // retries upon unpark of first thread
         boolean interrupted = false, first = false;
-        Node pred = null;                // predecessor of node when enqueued
+        Node pred = null, t;             // predecessor of node when enqueued
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedLongSynchronizer.java:296]: Nit: please don't use this style of multi-variable declaration as they are very easy to mis-read. Also the comment only applies to one of the variables.

File: src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedSynchronizer.java [hunk 2]
```diff
@@ -1553,6 +1610,25 @@ private void unlinkCancelledWaiters(ConditionNode node) {
...
+        private ConditionNode newConditionNode() {
+            int savedState;
+            if (tryInitializeHead() != null) {
+                try {
+                    return new ConditionNode();
+                } catch (OutOfMemoryError oome) {
+                }
+            }
+            if (!isHeldExclusively() || !release(savedState = getState()))
+                throw LockSupport.staticIllegalMonitorStateException; // OOM
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedSynchronizer.java:1626]: How is it possible to get IMSE this deep into the code? And the comment is confusing - OOM?

File: src/java.base/share/classes/java/util/concurrent/locks/LockSupport.java [hunk 3]
```diff
@@ -457,6 +457,46 @@ static final long getThreadId(Thread thread) {
         return thread.threadId();
     }
 
+    /**
+     *  Preallocated exceptions thrown if acquiring or releasing locks
+     *  when OutOfMemory.
+     */
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/util/concurrent/locks/LockSupport.java:463]: I don't see why this should be necessary. IMSE is thrown before any state changes occur and so it is is fine if the IMSE is replaced by OOME.

Even IE should be safe at the points it is thrown.

Also in both cases you want to see a full and proper stacktrace.

Comment (dholmes-ora) [reply src/java.base/share/classes/java/util/concurrent/locks/LockSupport.java:463]: Consider this another way, any place you have a `throw x` it must be safe to propagate that exception. It doesn't matter if that is actually `x` or an OOME caused by creating `x`.

**Comment by DougLea** [comment]
The idea here is that a non-OOME-throwing condition wait might be required for (some) memory to be reclaimed. This is not too farfetched for InterruptedException (if serving as a cancellation), so we allow it to be thrown (instead of OOME) in hopes that a catch of IE will free memory. It IS farfetched to think this might be the case for IllegalMonitorStateException, but I did it the same way for consistency.

File: src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedSynchronizer.java (hunk 2 above)
Comment (DougLea) [reply src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedSynchronizer.java:1626]: Clarified to:
            // fall through if encountered OutOfMemoryError
            if (!isHeldExclusively() || !release(savedState = getState()))
                throw LockSupport.staticIllegalMonitorStateException;

File: src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedLongSynchronizer.java (hunk 1 above)
Comment (DougLea) [reply src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedLongSynchronizer.java:296]: OK, changed.

File: test/jdk/ProblemList.txt [hunk 4]
```diff
@@ -495,7 +495,7 @@ java/lang/invoke/LFCaching/LFMultiThreadCachingTest.java        8151492 generic-
 java/lang/invoke/LFCaching/LFGarbageCollectedTest.java          8078602 generic-all
 java/lang/invoke/lambda/LambdaFileEncodingSerialization.java    8249079 linux-x64
 java/lang/invoke/RicochetTest.java                              8251969 generic-all
-java/lang/ref/OOMEInReferenceHandler.java                       8066859 generic-all
+java/lang/CompressExpandTest.java                               8287851 generic-all
```
Comment (AlanBateman) [inline test/jdk/ProblemList.txt:498]: There may be merge issue here. JDK-8287851 was fixed last week and the CompressExpandTest.java removed from the exclude list at the same time.

Comment (DougLea) [reply test/jdk/ProblemList.txt]: Thanks. fixed.

File: src/java.base/share/classes/java/util/concurrent/locks/LockSupport.java (hunk 3 above)
Comment (DougLea) [reply src/java.base/share/classes/java/util/concurrent/locks/LockSupport.java]: On further consideration, I agree. The very low probability that an IE or IMSE would enable further handling is not worth the extra startup cost of establishing static exceptions. Updated.

File: test/jdk/java/util/concurrent/locks/Lock/OOMEInAQS.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
+//package concurrent;
```
Comment (AlanBateman) [inline test/jdk/java/util/concurrent/locks/Lock/OOMEInAQS.java:1]: Do you mind adding a copyright header to the test?

Comment (DougLea) [reply test/jdk/java/util/concurrent/locks/Lock/OOMEInAQS.java]: OK, done.

**Review by AlanBateman — APPROVED** [review]

File: src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedLongSynchronizer.java [hunk 6]
```diff
@@ -1185,6 +1232,26 @@ private void unlinkCancelledWaiters(ConditionNode node) {
...
+            long savedState;
+            if (tryInitializeHead() != null) {
+                try {
+                    return new ConditionNode();
+                } catch (OutOfMemoryError oome) {
+                }
+            }
+            // fall through if encountered OutOfMemoryError
+            if (!isHeldExclusively() || !release(savedState = getState()))
+                throw new IllegalMonitorStateException();
```
Comment (dholmes-ora) [inline src/java.base/share/classes/java/util/concurrent/locks/AbstractQueuedLongSynchronizer.java:1249]: I still don't see how an OOME can possibly indicate we might somehow be in a state where we need to throw IMSE!

**Pull request closed without merging**